### ISSUE TITLE
datetime.now(timezone.utc) segfaults with Intel oneAPI v2022.1.2 and higher since Python 3.9 

### ISSUE THREAD

**Issue by haampie** [python/cpython#106424]
I've bisected this to 37fcbb65d4589fbb5a72153e9338cf8e6495f64f (first "bad" commit)

The following Python code is a minimal reproducer:

```python
from datetime import datetime, timezone
datetime.now(timezone.utc)
```

When compiling Python with any of the following Intel compilers:

```
2023.2.0
2023.1.0
2023.0.0
2022.2.1
```

Any version 3.9, 3.10, 3.11 causes a segfault in `add_datetime_timedelta` / `timezone_fromutc` where for some reason with `-O1` optimizations (edit: and higher) the `self->offset` / `delta` pointer is a `NULL`.

```c
Program received signal SIGSEGV, Segmentation fault.
0x00007f303942ef0c in add_datetime_timedelta (date=0x7f30394d7000, delta=0x0, factor=1) at Modules/_datetimemodule.c:5373
5373	    int day = GET_DAY(date) + GET_TD_DAYS(delta) * factor;
```

Adding `-fno-strict-aliasing` does not prevent this.

**Comment by corona10** [comment]
> Any version 3.9, 3.10, 3.11 causes a segfault in add_datetime_timedelta / timezone_fromutc where for some reason with -O1 optimizations the self->offset / delta pointer is a NULL.

From the original issue, -O2 doesn't occur any issue, Would you like to report to the Intel team first?
It can be a compiler bug.

**Comment by haampie** [comment]
It happens for -O1, -O2, -O3, it does not happen with -O0.

So yes, will try to bisect the intel compiler versions now... :weary:

**Comment by corona10** [comment]
> It happens for -O1, -O2, -O3, it does not happen with -O0.

Ah, thank you for the clarification.

**Comment by haampie** [comment]
When building Python 3.9.17 it segfaults with the following compilers:

- 2023.1.0
- 2023.0.0
- 2022.2.1

And succeeds with these:

- 2022.0.0
- 2021.1.2

So it could indeed be a compiler bug. Unclear how to show it is though.

**Comment by haampie** [comment]
Should I leave this issue open for discoverability? Others may run into it too. I've written a summary on the Spack package manager's repo https://github.com/spack/spack/pull/38710#issuecomment-1620517640, and we're hoping someone from Intel will confirm it's a compiler bug or not

**Comment by rscohn2** [comment]
@oleksandr-pavlyk: Are you building python with recent oneapi compilers?

**Comment by corona10** [comment]
@haampie cc @rscohn2 @oleksandr-pavlyk 

Would you like to check if the `-fno-strict-aliasing` flag is enabled in the compiler options?

**Comment by haampie** [comment]
It's not disabled:

```
checking whether /home/harmen/spack/lib/spack/env/oneapi/icx accepts and needs -fno-strict-aliasing... no
```

but when I build with `-fno-strict-aliasing` it still segfaults, so that's not the problem.

**Comment by haampie** [comment]
Hm, it looks like that configure check in Python is just there because some old GCC warned incorrectly about strict aliasing rules. (As in, it compiles some code with `-Werror -Wstrict-aliasing`, and if that fails, adds `-fno-strict-aliasing`, but any recent gcc / clang / icx would issue a warning).

In any case, the important thing is that `-fno-strict-aliasing` does not fix the segfault.

**Comment by vstinner** [comment]
For me, it's a compiler bug, I don't see how the C code could rely on an undefined behavior here. I suggest you reporting the bug to the Intel compiler.

**Comment by haampie** [comment]
@rscohn2 do you want to create a bug report for the compiler?

**Comment by rscohn2** [comment]
If you provide detailed information with clear evidence, the compiler group will respond quickly. Otherwise it will sit in the queue. This is because they get a lot of bogus bug reports that are application errors. 

I don't understand how everyone can be so sure it is a compiler bug. C programs have problems with uninitialized data, stale pointers. Optimizations change the layout of memory on stack and bugs and can be specific to compilers and releases.

There is another icx coming out next week. I will try that and if the error is still there then I will try to look at the bug.

**Comment by haampie** [comment]
It's already been said in this thread: the diff after which Intel Compilers produce a segfaulting Python is very small and innocent looking: https://github.com/python/cpython/commit/37fcbb65d4589fbb5a72153e9338cf8e6495f64f. It works fine on all major compilers. For a project like Python, I can only assume it uses sanitizers to detect UB and what not. If this was an "application error" I'm sure it would've surfaced elsewhere in the last 3 years.

**Comment by haampie** [comment]
Segfaults with oneAPI 2023.2.0 compilers too

**Comment by haampie** [comment]
With `-fsanitize=address`

```
$ ./python 
Python 3.10.12 (main, Jul 17 2023, 21:44:21) [Clang 17.0.0 (icx 2023.2.0.20230622)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from datetime import datetime, timezone
>>> datetime.now(timezone.utc)
AddressSanitizer:DEADLYSIGNAL
=================================================================
==214622==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000018 (pc 0x7f50b5c44f8d bp 0x7f50b7678988 sp 0x7ffeed8e0e20 T0)
==214622==The signal is caused by a READ memory access.
==214622==Hint: address points to the zero page.
    #0 0x7f50b5c44f8d in timezone_fromutc _datetimemodule.c
    #1 0x7f50ba502eb5 in cfunction_vectorcall_O methodobject.c
    #2 0x7f50ba4ba7bb in _PyObject_CallFunctionVa call.c
    #3 0x7f50ba4badea in _PyObject_CallMethodId (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0xbadea)
    #4 0x7f50b5c48d2f in datetime_datetime_now _datetimemodule.c
    #5 0x7f50ba502c89 in cfunction_vectorcall_FASTCALL_KEYWORDS methodobject.c
    #6 0x7f50ba59e900 in call_function ceval.c
    #7 0x7f50ba59ae00 in _PyEval_EvalFrameDefault (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x19ae00)
    #8 0x7f50ba593fa9 in _PyEval_Vector (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x193fa9)
    #9 0x7f50ba593f0d in PyEval_EvalCode (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x193f0d)
    #10 0x7f50ba5e6b0d in run_mod pythonrun.c
    #11 0x7f50ba5e5a26 in PyRun_InteractiveOneObjectEx pythonrun.c
    #12 0x7f50ba5e510d in _PyRun_InteractiveLoopObject (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x1e510d)
    #13 0x7f50ba5e4feb in _PyRun_AnyFileObject (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x1e4feb)
    #14 0x7f50ba5e57b3 in PyRun_AnyFileExFlags (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x1e57b3)
    #15 0x7f50ba6056e3 in Py_RunMain (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x2056e3)
    #16 0x7f50ba60589f in pymain_main main.c
    #17 0x7f50ba6058eb in Py_BytesMain (/home/harmen/spack/opt/spack/linux-ubuntu23.04-zen2/oneapi-2023.2.0/python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/lib/libpython3.10.so.1.0+0x2058eb)
    #18 0x7f50ba023a8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #19 0x7f50ba023b48 in __libc_start_main csu/../csu/libc-start.c:360:3
    #20 0x41f724 in _start (/tmp/harmen/spack-stage/spack-stage-python-3.10.12-rp4tihyd4czddiqmgydgrz4mkrkebbz2/spack-src/python+0x41f724)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV _datetimemodule.c in timezone_fromutc
```

other sanitizers I can't get to work with icx.

**Comment by rscohn2** [comment]
I spent an afternoon narrowing down the problem to the store for this line not happening: https://github.com/python/cpython/blob/a293fa5915c21b21f5cb8ed9649fbdb37b4c1421/Modules/_datetimemodule.c#L1165
I added a printf to show that Py_NewRef(offset);  returns it's input parameter, but the value is not stored info self->offset.
I filed a ticket with the compiler that shows the asm output for the function and why I believe the store is missing. It does look like a compiler bug to me, but let's see what they say.
